### PR TITLE
feat(aurelia): single "aurelia" package to re-exports internals

### PR DESCRIPTION
**Pull request by 3cp**
# Pull Request

## 📖 Description

Re-export all in a single "aurelia" package, and a wrapper to start app.

### 🎫 Issues

This is an implementation of RFC #630, more description in the RFC.

## 👩‍💻 Reviewer Notes

Our release and publish scripts seems ok to handle non-scoped package. I only did a minor update.

The re-exports only exported essentials from internal packages, lots of exports are commented out right now.

**Please review and make modification to the export list, especially on kernel, router, and runtime.**

Note fetch-client and router are in the re-exports list, as I think at least 90% of Aurelia 2 users will use them.

i18n, dialog, validation, store, are not in the list. We can change in future if needed.

The quick start wrapper is the default export of "aurelia" package. It followed @EisenbergEffect 's suggested API.

```ts
import Aurelia, { StyleConfiguration, RouterConfiguration } from 'aurelia';
import { MyRootComponent } from './my-root-component';
// By default host to element name (<my-root-component> for MyRootComponent),
// or <body> if <my-root-component> is absent.
Aurelia.app(MyRootComponent).start();

// Or load additional aurelia features
Aurelia
  .register(
    StyleConfiguration.shadowDOM(),
    RouterConfiguration.customize({ useUrlFragmentHash: false })
  )
  .app(MyRootComponent)
  .start();

// Or host to <my-start-tag>
Aurelia
  .register(
    StyleConfiguration.shadowDOM(),
    RouterConfiguration.customize({ useUrlFragmentHash: false })
  )
  .app({
    component: MyRootComponent,
    host: document.querySelector('my-start-tag')
  })
  .start();
```

The method returns Aurelia instance, not the task. User can wait on it. `au(MyApp).wait().then(() => ...);`.

## 📑 Test Plan

Tested manually in a local webpack app. The quick start wrapper worked as expected.

Similiar to existing Aurelia class, there is no unit tests for the quick start wrapper. I don't know how to do it. However it will be covered by e2e tests with various example apps.

## ⏭ Next Steps

- [ ] update conventions. (Let's do it in a separate PR.)
- [ ] update all example apps setup.
- [ ] update aurelia/new makes skeletons.
- [ ] update docs.

### CODE REVIEW

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://28434-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by bigopon** [comment]
I think it looks good to start with. Lgtm

File: packages/aurelia/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,923 @@
...
+  // ISubscriber,
+  // isIndexMap,
+  // copyIndexMap,
+  // cloneIndexMap,
+  // createIndexMap,
+
+  // instructionRenderer,
+  // ensureExpression,
+  // addComponent,
+  // addBinding,
```
Comment (fkleuver) [inline packages/aurelia/src/index.ts]: Should export these 4, surface api decorator + 3 crucial helpers

File: packages/aurelia/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,923 @@
...
+
+  // AttributeBindingRendererRegistration,
+  // ListenerBindingRendererRegistration,
+  // SetAttributeRendererRegistration,
+  // StylePropertyBindingRendererRegistration,
+  // TextBindingRendererRegistration,
+
+  // DefaultRenderers,
+
+  BasicConfiguration as RuntimeHtmlBasicConfiguration,
```
Comment (fkleuver) [inline packages/aurelia/src/index.ts]: No need to export this 1

File: packages/aurelia/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,923 @@
...
+
+  // createElement,
```
Comment (fkleuver) [inline packages/aurelia/src/index.ts]: Should export this 1. surface api

File: packages/aurelia/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,923 @@
...
+  // AdoptedStyleSheetsStyles,
+  // StyleElementStyles,
+  // IShadowDOMStyles,
+  // IShadowDOMGlobalStyles
+} from '@aurelia/runtime-html';
+
+export {
+  // IDOMInitializerRegistration,
+  // DefaultComponents as RuntimeHtmlBrowserDefaultComponents,
+  BasicConfiguration as RuntimeHtmlBrowserBasicConfiguration
```
Comment (fkleuver) [inline packages/aurelia/src/index.ts]: No need to export this 1

File: packages/aurelia/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,923 @@
...
+
+  containerless,
+  customElement,
+  CustomElementHost,
+  CustomElement,
+  ICustomElementDecorator,
+  ICustomElementResource,
+  ICustomElementType,
+  IElementProjector,
+  IProjectorLocator,
```
[the remark this answers is not in the document]
Comment (3cp) [reply packages/aurelia/src/index.ts]: what about the following types for value converter?
  IValueConverter,
  IValueConverterDefinition,
  IValueConverterResource,
  IValueConverterType

File: packages/aurelia/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,923 @@
...
+  RecognizeResult,
+  RecognizeResults,
+  CharSpec,
+  // State as RouterState, // duplicated in @aurelia/runtime
+  StaticSegment,
+  DynamicSegment,
+  StarSegment,
+  EpsilonSegment,
+  Segment,
+  RouteRecognizer,
```
[the remark this answers is not in the document]
Comment (3cp) [reply packages/aurelia/src/index.ts]: Where is no typing needed for something like LinkHandler, Nav, NavRoute, RouteHandler?

For example, user of vcurrent can extend DefaultLinkHandler and replace it in DI.

File: packages/aurelia/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,923 @@
...
+  // Writable,
+  // IfEquals,
+  // ReadonlyKeys,
+  // WritableKeys,
+
+  relativeToFile,
+  join,
+  buildQueryString,
+  parseQueryString,
+  IQueryParams,
```
[the remark this answers is not in the document]
Comment (3cp) [reply packages/aurelia/src/index.ts]: Might be useful for plugin authors.

File: packages/aurelia/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,923 @@
...
+export {
+  Interceptor,
+  RetryConfiguration,
+  RetryableRequest,
+  ValidInterceptorMethodName,
+
+  json,
+
+  retryStrategy,
+  RetryInterceptor,
```
[the remark this answers is not in the document]
Comment (3cp) [reply packages/aurelia/src/index.ts]: json is a surface API in vcurrent. It was used in some tutorial I remember.

**Comment by 3cp** [comment]
Exports list is updated as requested. Left few comments for discussion.

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://28535-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by 3cp** [comment]
@fkleuver `test/realworld` app is using following exports `json`, `INavRoute`, `IObserverLocator`, `IViewModel`, `NavRoute`.

These cherry-picking is troublesome, why not just re-export as many as we can?

File: packages/aurelia/src/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (fkleuver) [reply packages/aurelia/src/index.ts]: That's what the interfaces are for. A few of them are still msising

**Comment by fkleuver** [comment]
@3cp 

> test/realworld app is using following exports json, INavRoute, IObserverLocator, IViewModel, NavRoute.

We can export those, no problem.

> These cherry-picking is troublesome, why not just re-export as many as we can?

Either we do the cherry-picking or we leave the end user to weed through 250 exports they'll never use 😄

**Comment by 3cp** [comment]
It seems `IViewModel` can be removed from realworld app. @jwx do you want me to update realworld app to latest code (simplified import from "aurelia")? I can leave that to you too.

**Comment by jwx** [comment]
@3cp Leave it, please. I've got changes coming up in a PR and it'll just make rebase complain.

**Pull request merged**